### PR TITLE
migration gude: python 2 => 3 implications

### PR DESCRIPTION
**Pull request by oliver-sanders**
Add a major changes page covering Python changes which may need to be made during Cylc 7 -> 8 porting.

This is a list off the top of my head, can we think of anything else?

### CODE REVIEW

**Comment by oliver-sanders** [comment]
(note this requires a Rose target which was only just added https://github.com/metomi/rose/pull/2563)

File: src/7-to-8/major-changes/python-2-3.rst [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+Cylc 7 ran under Python 2, Cylc 8 runs under Python 3.
+
+Cylc can be extended with custom Python scripts. These scripts are run under
+the same version of Python used by Cylc.
+
+As a result if you are moving from Cylc 7 to Cylc 8 you must upgrade any
+scripts from Python 2 to Python 3 in the process.
+
+If you want to support both Cylc 7 and 8 you must support both Python 2 and 3.
+There are tools to help you do this e.g. `six`_.
```
Comment (wxtim) [inline src/7-to-8/major-changes/python-2-3.rst:28]:
```suggestion
There are tools such as `six`_ to help you do this.
```

File: src/7-to-8/major-changes/python-2-3.rst [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+.. rubric:: Convert Python code from Cylc 7 to Cylc 8:
+
+.. code-block:: diff
+
+   - from cylc import LOG
+   + from cylc.flow import LOG
+   - from isodatetime.data import Duration
+   + from metomi.isodatetime.data import Duration
+
+.. rubric:: Pyhton code which supports both Cylc 7 & Cylc 8:
```
Comment (wxtim) [inline src/7-to-8/major-changes/python-2-3.rst:78]:
```suggestion
.. rubric:: Python code which supports both Cylc 7 & Cylc 8:
```

**Review by datamel — COMMENTED** [review]
I am getting a build error locally : `WARNING: undefined label: rose:builtin.rose_ana`
I presume this is due to the intersphinx mapping `http://metomi.github.io/rose/doc/html` not having updated with the required change?

**Comment by oliver-sanders** [comment]
Yes, tests will fail until metomi-rose 2.0rc2 docs are published.

File: src/7-to-8/major-changes/python-2-3.rst [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+      {% from "os" import path %}
+
+:ref:`Custom Trigger Functions`
+   Any custom x-trigger functions.
+
+
+Module Changes
+--------------
+
+Three Python modules have been renamed between Cylc 7 and Cylc 8:
```
Comment (hjoliver) [inline src/7-to-8/major-changes/python-2-3.rst:59]:
```suggestion
Three Python packages have been renamed between Cylc 7 and Cylc 8:
```

File: src/7-to-8/major-changes/python-2-3.rst [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+* ``cylc`` => ``cylc.flow``
+* ``isodatetime`` => ``metomi.isodatetime``
+* ``rose`` => ``metomi.rose``
+
+If you are importing these modules you will need to update the module names.
```
Comment (hjoliver) [inline src/7-to-8/major-changes/python-2-3.rst:65]:
```suggestion
If you are importing from these packages you will need to update the package names.
```

**Review by hjoliver — APPROVED** [review]
All good apart from `module` -> `package`.

File: src/7-to-8/major-changes/python-2-3.rst [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+   .. code-block:: jinja
+
+      {% from "os" import path %}
+
+:ref:`Custom Trigger Functions`
+   Any custom x-trigger functions.
+
+
+Module Changes
+--------------
```
Comment (oliver-sanders) [inline src/7-to-8/major-changes/python-2-3.rst:57]:
```suggestion
Package Name Changes
--------------------
```

**Review by wxtim — APPROVED** [review]
Happy with response to review

**Comment by oliver-sanders** [comment]
Pinning Rose via intersphinx to RC1, will need to keep this updated until Rose 2.0 is released (at which point it will become the default documented version)

**Comment by datamel** [comment]
The rose pinning has fixed the rose-ana link error. Thanks. Possible problem locally now: 
```
~/cylc-doc/src/glossary.rst:1099: WARNING: undefined label: restarting suites
~/cylc-doc/src/glossary.rst:1079: WARNING: undefined label: stopping suites
<unknown>:1492: WARNING: term not in glossary: wall-clock time`

**Comment by oliver-sanders** [comment]
Hmm, I think these glossary items were being pulled in from the Rose docs, they need updating to the cylc loactions.

**Comment by oliver-sanders** [comment]
Fixed two of the issues, however, the third requires a change to cylc-flow.

```diff
diff --git a/cylc/flow/cfgspec/workflow.py b/cylc/flow/cfgspec/workflow.py
index c254fd597..6732adad6 100644
--- a/cylc/flow/cfgspec/workflow.py
+++ b/cylc/flow/cfgspec/workflow.py
@@ -1103,7 +1103,7 @@ with Conf(
                 behavior between submission attempts.
             ''')
             Conf('execution time limit', VDR.V_INTERVAL, desc='''
-                Set the execution (:term:`wall-clock <wall-clock time>`) time
+                Set the execution (:term:`wallclock <wallclock time>`) time
                 limit of a task job.
 
                 For ``background`` and ``at`` job runners Cylc invokes the
```

**Comment by oliver-sanders** [comment]
https://github.com/cylc/cylc-flow/pull/4775

**Comment by oliver-sanders** [comment]
In the mean time (to allow release) I'll add a duplicate glossary definition...

**Comment by datamel** [comment]
> In the mean time (to allow release) I'll add a duplicate glossary definition...

Good thinking!

File: src/7-to-8/major-changes/python-2-3.rst [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+   custom Python scripts.
+
+   This does not impact :term:`task` scripts, Cylc can still run Python 2
+   tasks if desired.
+
+
+Overview
+--------
+
+.. _six: src/7-to-8/major-changes/python-2-3.py
```
Comment (datamel) [inline src/7-to-8/major-changes/python-2-3.rst:17]: Not sure this should be a .py?
```suggestion
.. _six: src/7-to-8/major-changes/python-2-3.rst
```

Comment (oliver-sanders) [reply src/7-to-8/major-changes/python-2-3.rst:17]: Yikes, more like:

```suggestion
.. _six: https://pypi.org/project/six/
```

**Review by datamel — APPROVED** [review]
Builds nicely, thanks @oliver-sanders

**Pull request merged**